File: frontend/src/api/graph.ts
import { apiClient } from "./client";
import type { Subgraph } from "./types";

export interface SubgraphQuery {
  projectId: string;
  rootId: string;
  depth: number;
  edgeTypes?: string[];
}

export const graphApi = {
  subgraph: ({ projectId, rootId, depth, edgeTypes }: SubgraphQuery) =>
    apiClient.get<Subgraph>(`/api/projects/${projectId}/graph/subgraph`, {
      root_id: rootId,
      depth,
      edge_type: edgeTypes,
    }),
};

File: backend/src/loregraph/api/routers/graph.py
from typing import Annotated

from fastapi import APIRouter, Query

from loregraph.api.deps import EdgeStoreDep, EntityStoreDep
from loregraph.schemas.graph import SubgraphOut
from loregraph.services.graph_query import get_subgraph

router = APIRouter(prefix="/projects/{project_id}/graph", tags=["graph"])


@router.get("/subgraph", response_model=SubgraphOut)
async def subgraph(
    project_id: str,
    entity_store: EntityStoreDep,
    edge_store: EdgeStoreDep,
    root_id: str,
    depth: Annotated[int, Query(ge=0, le=10)] = 1,
    edge_type: Annotated[list[str] | None, Query()] = None,
) -> SubgraphOut:
    edge_types = frozenset(edge_type) if edge_type else None
    return await get_subgraph(
        entity_store, edge_store, project_id, root_id, depth, edge_types
    )
